src/main/java/org/mariadb/jdbc/HostAddress.java

```diff
@@ -107,11 +107,16 @@ public HostAddress(String host, int port, String type) {
         String[] tokens = spec.trim().split(",");
         List<HostAddress> arr = new ArrayList<>(tokens.length);
 
-        for (int i = 0; i < tokens.length; i++) {
-            if (tokens[i].startsWith("address=")) {
-                arr.add(parseParameterHostAddress(tokens[i]));
+        // Allow only cluster endpoint to be given unless testing
+        if (haMode == HaMode.AURORA && (tokens.length > 1 || (tokens.length == 1 && spec.indexOf(".cluster-") < 0)) && System.getProperty("auroraFailoverTesting") == null) {
+            throw new IllegalArgumentException("Invalid Aurora connection URL, address should correspond to cluster endpoint");
```

Comment:
But keeping the current single instance implementation in Utils would not allow auto-discovery since the self-discovery feature only applies to Aurora specific classes (listener, protocol), more specifically, the code to allow auto-discovery is in AuroraListener and not MastersFailoverListener which is the listener that is instantiated for single endpoint.